substrate/frame/opf/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+
+		/// Project added to whitelisted projects list
+		Projectlisted { when: ProvidedBlockNumberFor<T>, project_id: ProjectId<T> },
+
+		/// Project removed from whitelisted projects list
+		ProjectUnlisted { when: ProvidedBlockNumberFor<T>, project_id: ProjectId<T> },
+
+		/// Project Funding Accepted by voters
+		ProjectFundingAccepted {
+			project_id: ProjectId<T>,
+			when: ProvidedBlockNumberFor<T>,
+			round_number: u32,
+			amount: BalanceOf<T>,
+		},
+
+		/// Reward claim has expired
+		ExpiredClaim { expired_when: ProvidedBlockNumberFor<T>, project_id: ProjectId<T> },
+
+		/// Project Funding rejected by voters
+		ProjectFundingRejected { when: ProvidedBlockNumberFor<T>, project_id: ProjectId<T> },
+
+		/// A new voting round started
+		VotingRoundStarted { when: ProvidedBlockNumberFor<T>, round_number: u32 },
+
+		/// The users voting period ended. Reward calculation will start.
+		VoteActionLocked { when: ProvidedBlockNumberFor<T>, round_number: u32 },
+
+		/// The voting round ended
+		VotingRoundEnded { when: ProvidedBlockNumberFor<T>, round_number: u32 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Not enough Funds in the Pot
+		InsufficientPotReserves,
+		/// The funds transfer operation failed
+		TransferFailed,
+		/// Spend or Spend index does not exists
+		InexistentSpend,
+		/// No valid Account_id found
+		NoValidAccount,
+		/// No project available for funding
+		NoProjectAvailable,
+		/// The Funds transfer failed
+		FailedSpendOperation,
+		/// Still not in claiming period
+		NotClaimingPeriod,
+		/// Funds locking failed
+		FundsReserveFailed,
+		/// An invalid result  was returned
+		InvalidResult,
+		/// The reward calculation failed due to an internal error
+		FailedRewardCalculation,
+		/// Voting round is over
+		VotingRoundOver,
+		/// This voting round does not exists
+		NoRoundFound,
+		/// Maximum number of projects submission for reward distribution as been reached
+		MaximumProjectsNumber,
+		/// Another project has already been submitted under the same project_id
+		SubmittedProjectId,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<SystemBlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_n: SystemBlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			Self::on_idle_function(remaining_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// OPF Projects registration
+		///
+		/// ## Dispatch Origin
+		///
+		/// Must be signed
+		///
+		/// ## Details
+		///
+		/// From this extrinsic only Root can register project.
+		///
+		/// ### Parameters
+		/// - `project_id`: The account that will receive the reward.
+		///
+		/// ### Errors
+		/// - [`Error::<T>::MaximumProjectsNumber`]: Maximum number of project subscriptions reached
+		///  
+		/// ## Events
+		/// Emits [`Event::<T>::Projectlisted`].
+		#[pallet::call_index(0)]
+		#[transactional]
+		pub fn register_project(origin: OriginFor<T>, project_id: ProjectId<T>) -> DispatchResult {
+			let _caller = ensure_root(origin)?;
```

Comment:
Should be a custom origin configured in the `Config` trait. Something like `AdminOrigin` or whatever.